relayer/chains/substrate/event_parser.go

```diff
@@ -1,16 +1,105 @@
 package substrate
 
 import (
+	"fmt"
+	"strconv"
+
 	rpcclienttypes "github.com/ComposableFi/go-substrate-rpc-client/v4/types"
+	beefyclienttypes "github.com/ComposableFi/ics11-beefy/types"
 	clienttypes "github.com/cosmos/ibc-go/v5/modules/core/02-client/types"
-	"github.com/cosmos/relayer/v2/relayer/processor"
 	"github.com/cosmos/relayer/v2/relayer/provider"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
 )
 
+// ibcMessage is the type used for parsing all possible properties of IBC messages
+type ibcMessage struct {
+	eventType string
+	info      ibcMessageInfo
+}
+
+type ibcMessageInfo interface {
+	parseAttrs(log *zap.Logger, attrs interface{})
+	MarshalLogObject(enc zapcore.ObjectEncoder) error
+}
+
 // ibcMessagesFromTransaction parses all events within a transaction to find IBC messages
-func (scp *SubstrateChainProcessor) handleIBCMessagesFromEvents(ibcEvents rpcclienttypes.IBCEventsQueryResult, height uint64, c processor.IBCMessagesCache) error {
-	//TODO implement me
-	panic("implement me")
+func (scp *SubstrateChainProcessor) ibcMessagesFromTransaction(ibcEvents rpcclienttypes.IBCEventsQueryResult, height uint64) (messages []ibcMessage) {
+	packetAccumulator := make(map[uint64]*packetInfo)
```

Comment:
Can we expect that all of the `ibcEvents` are relative to a single channel? If not, I think the key here will need to change from `uint64` to a type that has the sequence and the channel information, e.g.:
```golang
type ibcPacketKey struct {
  sequence uint64
  channelKey provider.ChannelKey
}
```

Reply:
Added